hudi-client/hudi-flink-client/src/main/java/org/apache/hudi/io/v2/RowDataLogWriteHandle.java

```diff
@@ -0,0 +1,366 @@
+        FSUtils.constructAbsolutePath(config.getBasePath(), partitionPath),
+        hoodieTable.getPartitionMetafileFormat());
+    partitionMetadata.trySave();
+  }
+
+  private void initWriteConf(StorageConfiguration<?> storageConf, HoodieWriteConfig writeConfig) {
+    storageConf.set(
+        HoodieStorageConfig.PARQUET_WRITE_UTC_TIMEZONE.key(),
+        writeConfig.getString(HoodieStorageConfig.PARQUET_WRITE_UTC_TIMEZONE.key()));
+  }
+
+  private MetadataValues populateMetadataFields(HoodieRecord<T> hoodieRecord) {
+    MetadataValues metadataValues = new MetadataValues();
+    if (config.populateMetaFields()) {
+      String seqId =
+          HoodieRecord.generateSequenceId(instantTime, getPartitionId(), RECORD_COUNTER.getAndIncrement());
+      metadataValues.setFileName(fileId);
+      metadataValues.setPartitionPath(partitionPath);
+      metadataValues.setRecordKey(hoodieRecord.getRecordKey());
+      metadataValues.setCommitTime(instantTime);
+      metadataValues.setCommitSeqno(seqId);
+    }
+    if (config.allowOperationMetadataField()) {
+      metadataValues.setOperation(hoodieRecord.getOperation().getName());
+    }
+    return metadataValues;
+  }
+
+  private void processAppendResult(AppendResult result, HoodieLogBlock dataBlock) {
+    HoodieDeltaWriteStat stat = (HoodieDeltaWriteStat) this.writeStatus.getStat();
+
+    Preconditions.checkArgument(stat.getPath() == null, "Only one append are expected for " + this.getClass().getSimpleName());
+    // first time writing to this log block.
+    updateWriteStatus(stat, result);
+
+    // for parquet data block, we can get column stats from parquet footer directly.
+    if (config.isMetadataColumnStatsIndexEnabled()) {
+      Set<String> columnsToIndexSet = new HashSet<>(HoodieTableMetadataUtil
+          .getColumnsToIndex(hoodieTable.getMetaClient().getTableConfig(),
+              config.getMetadataConfig(), Lazy.eagerly(Option.of(writeSchemaWithMetaFields)),
+              Option.of(HoodieRecord.HoodieRecordType.FLINK)).keySet());
+
+      Map<String, HoodieColumnRangeMetadata<Comparable>> columnRangeMetadata;
+      if (dataBlock == null) {
+        // only delete block exists
+        columnRangeMetadata = new HashMap<>();
+        for (String col: columnsToIndexSet) {
+          columnRangeMetadata.put(col, HoodieColumnRangeMetadata.create(
+              stat.getPath(), col, null, null, 0L, 0L, 0L, 0L));
+        }
+      } else {
+        ValidationUtils.checkArgument(dataBlock instanceof ColumnRangeMetadataProvider,
+            "Log block for Flink ingestion should always be an instance of ColumnRangeMetadataProvider for collecting column stats efficiently.");
+        columnRangeMetadata =
+            ((ColumnRangeMetadataProvider) dataBlock).getColumnRangeMeta().entrySet().stream()
+                .filter(e -> columnsToIndexSet.contains(e.getKey()))
+                .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().copy(stat.getPath())));
```

Comment:
Why we must copy the file path?

Reply:
Column range meta is collected during data block serializing, and data block do not have the context about the file to write, so write handle here supplement file path field for column range meta.

Reply:
> Column range meta is collected during data block serializing

Can we do it in the footer -> metadata transformation?